napari/experimental/_progressive_loading.py

```diff
@@ -0,0 +1,1649 @@
+    chunk_map = chunk_maps[scale]
+    scale_factor = scale_factors[scale]
+
+    # Points for each chunk, for example, centers
+    points = np.array(list(chunk_map.keys()))
+
+    # Mask of whether points are within our interval, this is in array
+    # coordinates
+    point_mask = np.array(
+        [
+            np.all(point >= min_coord) and np.all(point <= max_coord)
+            for point in points
+        ]
+    )
+
+    # Rescale points to world for priority calculations
+    points_world = points * np.array(scale_factor)
+
+    # Prioritize chunks using world coordinates
+    distances = distance_from_camera_center_line(points_world, camera)
+    depth = visual_depth(points_world, camera)
+    priorities = prioritized_chunk_loading_3D(
+        depth, distances, camera.zoom, alpha=alpha, visible=point_mask
+    )
+
+    # Select the number of chunks
+    # TODO consider using threshold on priorities
+    """
+    Note:
+    switching from recursing on 1 top chunk to N-best introduces extra
+    complexity, because the shape of texture allocation needs to
+    accommodate projections from all viewpoints around the volume.
+    """
+    n = 1
+    best_priorities = np.argsort(priorities)[:n]
+
+    # Iterate over points/chunks and add corresponding nodes when appropriate
+    for idx, point in enumerate(points):
+        # TODO: There are 2 strategies here:
+        # 1. Render *visible* chunks, or all if we're on the last scale level
+        #    Skip the chunk at this resolution because it will be shown in
+        #    higher res. This fetches less data.
+        # if point_mask[idx] and (idx not in best_priorities or scale == 0):
+        # 2. Render all chunks because we know we will zero out this data when
+        #    it is loaded at the next resolution level.
+        if point_mask[idx]:
+            coord = tuple(point)
+            chunk_slice = chunk_map[coord]
+            offset = [sl.start for sl in chunk_slice]
+
+            # When we get_chunk chunk_slice needs to be in data space, but
+            # chunk slices are 3D
+            data_slice = tuple(
+                [slice(el, el + 1) for el in dims.current_step[:-3]]
+                + [slice(sl.start, sl.stop) for sl in chunk_slice]
+            )
+
+            data = get_chunk(
+                data_slice,
+                array=array,
+                dtype=dtype,
+            )
+
+            # Texture slice (needs to be in layer.data dimensions)
+            # TODO there is a 3D ordering assumption here
+            texture_slice = tuple(
+                [slice(el, el + 1) for el in dims.current_step[:-3]]
+                + [
+                    slice(sl.start - offset, sl.stop - offset)
+                    for sl, offset in zip(chunk_slice, min_coord)
+                ]
+            )
+            if texture_slice[1].start < 0:
+                import pdb
+
+                pdb.set_trace()
+
+            # TODO consider a data class instead of a tuple
```

Comment:
yeah, it's pretty unclear what any of these mean, especially with no docstring. also im generally in favour of namedtuples instead